liquigraph-core/src/main/java/org/liquigraph/core/io/lock/LockableConnection.java

```diff
@@ -0,0 +1,129 @@
+ * limitations under the License.
+ */
+package org.liquigraph.core.io.lock;
+
+import org.liquigraph.core.exception.LiquigraphLockException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.UUID;
+
+public final class LockableConnection implements ConnectionSupplier {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LockableConnection.class);
+    private final Connection connection;
+    private final UUID uuid;
+    private final Thread task;
+
+    public LockableConnection(Connection connection) {
+        this.connection = connection;
+        this.uuid = UUID.randomUUID();
+        this.task = new Thread(new ShutdownTask());
+        initiateLocking();
+    }
+
+    private final void initiateLocking() {
+        registerLockRemovalHook();
+        ensureLockUnicity();
+        tryWriteLock();
+    }
+
+   /**
+    * Closes the underlying connection
+    */
+   @Override
+    public void close() {
+        Runtime.getRuntime().removeShutdownHook(task);
+        closeConnection();
+    }
+
+    @Override
+    public Connection get() {
+        return connection;
+    }
+
+    private final void registerLockRemovalHook() {
+        Runtime.getRuntime().addShutdownHook(task);
+    }
+
+    private final void ensureLockUnicity() {
+        try (Statement statement = connection.createStatement()) {
+            statement.execute("CREATE CONSTRAINT ON (lock:__LiquigraphLock) ASSERT lock.name IS UNIQUE");
+            connection.commit();
+        }
+        catch (SQLException e) {
+            throw new LiquigraphLockException(
+                "Could not ensure __LiquigraphLock unicity\n\t" +
+                    "Please make sure your instance is in a clean state\n\t" +
+                    "No more than 1 lock should be there simultaneously!",
+                e
+            );
+        }
+    }
+
+    private final void tryWriteLock() {
+        try (PreparedStatement statement = connection.prepareStatement(
+            "CREATE (:__LiquigraphLock {name:'John', uuid:{1}})")) {
+
+            statement.setString(1, uuid.toString());
+            statement.execute();
+            connection.commit();
+        }
+        catch (SQLException e) {
+            throw new LiquigraphLockException(
+                "Cannot create __LiquigraphLock lock\n\t" +
+                    "Likely another Liquigraph execution is going on or has crashed.",
+                e
+            );
+        }
+    }
+
+    private final void closeConnection() {
+        try {
+            if (!connection.isClosed()) {
+                task.run();
+                connection.close();
+            }
+        }
+        catch (SQLException e) {
+            LOGGER.error("Could not release connection.", e);
+        }
+    }
+
+    private final class ShutdownTask implements Runnable {
```

Comment:
I'd use a nested class instead of an inner class, simply sharing the `uuid` property instead of the whole `LockableConnection`, to limit the object graph seen and held by the hook. It's probably a question of taste, but I avoid nested classes whenever I can, especially when they escape the scope of the enclosing class.

Who knows, there might be a bug lurking in `removeShutdownHook()`? :)